### DISCUSSION TITLE
Current State, Future Vision for Fabric.js

### DISCUSSION THREAD

**Discussion by usamshen** [fabricjs/fabric.js#9993] [General]
Hi Fabric.js Community,

I am facing challenges in understanding the full picture of the current state and the future vision of Fabric.js, including the challenges we face and the great ideas we have. To facilitate better understanding and contributions, I propose creating visualizations to share the big picture. If we could create a shared vision, it can also help to drive the energy from the community to better work together.

hope to get feedback from @andrea11 @ShaMan123 @jiayihu  and the rest of the community...

**Visualizing the Current State:**
I've attached a visualization of Fabric.js's current state, including major classes, What are the known important issues? Are there any areas that need more clarity? Are we missing any critical components or issues? Are there any great PRs there pending on merging?
![image](https://github.com/user-attachments/assets/026e7e1a-5e0f-44f8-aa84-eeb79a5a1355)

**Envisioning the Future State:**
what the future state of Fabric.js should look like. What features, improvements, and innovations should we aim for? How do we envision the user experience evolving? thoughts, ideas, and visions for the future?
 
The collaborative board is created at: https://app.boardx.us/invite/eAgpvy. 
I am using Fabric.js to build a digital open-source whiteboard solution called BoardX which integrates with GenAI. Welcome to use BoardX for Fabricjs community projects and provides us feedback. Thanks

**Comment by asturur** [comment]
@usamshen the topic is so broad.

There is a glaring issue with documentation and availablity of good examples, and that is the area where i'm spending most of my time.
All areas needs more clarity, without documentation is also hard to say where are we, and writing documentation sometimes you realize things are in unoptimal state and need to change.
Not having documentation is also the reason for 75% of the opened issues, that are a large timesink.

Feature wise the most important and the next feature i would like to see done are partial re-renders ( performance ) and support for RTL text ( availability to more people/cultures ).

For what regards the API, i think something that requires attention is how the developer can build it's own interactive flow.
The Controls api helped a lot but something more flexible is needed to handle:

- drag selection
- drag
- multi selection
- seleciton
- painting
- brushes

When a developer wants to implement something like drawing a polygon by hand, or connect 2 objects with a connector he/she needs to go through some loops to avoid default fabricJS hardcoded interactions.

Technology wise fabric caching needs to be improved.
Sometimes we need context isolation without the current implementation of caching and that requires a lot of custom work.
Disabling caching is needed for high quality exports but caching is also needed for a series of effects that you may need in an export.

Regarding classes and that schema you have drawn, that long column of Object classes should be squished as much as possible, Line, Polyline, Polygon and Path should loose their specific code and be rebased on Path. Text, Itext and Textbox should become just a single thing, a Text class with a wrapping strategy, and an optional interactivity manager that you can attach to it. But all those changes are annoying when in need of an update and are not a game changer ever. Is an incremental polishing game that needs to be balanced with stability and new features in my opinion.

There are many many things that have been thought through, none of them are game changers in comparison to documentation in my opinion.

**Reply by usamshen** [reply]
it would be great to know what you were thinking and maybe ideas from others, and then build a full picture...and i used ai to summarize the thoughts, most of the items are needed for boardx. I added these items to the board. 

1. **Documentation Improvements**:
   - Comprehensive and clear documentation.
   - Examples and use cases.
   - Documentation updates to reflect current state and changes.

2. **Performance Enhancements**:
   - Partial re-renders to improve performance.

3. **Text Support**:
   - Support for RTL (Right-to-Left) text to increase accessibility for more cultures.

4. **API Flexibility**:
   - More flexible APIs for custom interactive flows.
   - Enhanced control APIs to handle:
     - Drag selection
     - Drag
     - Multi-selection
     - Selection
     - Painting
     - Brushes
   - Simplify custom interactions such as drawing polygons or connecting objects.

5. **Technology and Caching**:
   - Improved caching mechanisms.
   - Context isolation without complex custom work.
   - Options to disable caching for high-quality exports while maintaining needed effects.

6. **Class Structure Optimization**:
   - Consolidate Object classes:
     - Rebase Line, Polyline, Polygon, and Path on Path.
     - Merge Text, IText, and Textbox into a single Text class with a wrapping strategy and optional interactivity manager.
   - Balance between incremental improvements and stability/new features.

**Comment by usamshen** [comment]
i bet there are shared needs and not shared needs
![image](https://github.com/user-attachments/assets/e08cb59e-b8e5-4044-a420-a70378fe6ee8)

Further idea is that we could do some planning activities, brainstorm all the features on the roadmap from the maintainer and requirements, ideas, PoCs from the community contributors, maintainer could prioritize the needs. 
![image](https://github.com/user-attachments/assets/6f988be9-e440-4efb-b043-1a84994f3cd3)

**Comment by usamshen** [comment]
Here are some requirements from BoardX. While some of these features are already implemented in Fabric.js V6, they have not yet been integrated into BoardX. If a feature is shared with Fabric.js, we should add it to the Fabric.js development list. Otherwise, it will be included in the BoardX (we call the widgets CanvasX, it is under MIT license) feature list.


![image](https://github.com/user-attachments/assets/449f7c20-f596-4ff8-9895-556e4aed1fd4)
